### PR TITLE
Allow build & run to reference unreachable targets

### PR DESCRIPTION
**Pull request by hartbit**

### CODE REVIEW

File: Sources/Build/BuildPlan.swift [hunk 1]
```diff
@@ -474,7 +474,7 @@ public class BuildPlan {
 
         // Create build target description for each target which we need to plan.
         var targetMap = [ResolvedTarget: TargetDescription]()
-        for target in graph.targets {
+        for target in graph.allTargets {
```
Comment (aciidb0mb3r) [inline Sources/Build/BuildPlan.swift]: Won't this start building all the targets when `swift build` is invoked?

File: Sources/Commands/SwiftRunTool.swift [hunk 2]
```diff
@@ -126,27 +126,30 @@ public class SwiftRunTool: SwiftTool<RunToolOptions> {
...
         if let executable = options.executable {
-            // If the exectuable is explicitly specified, verify that it exists.
-            guard let executableProduct = executableProducts.first(where: { $0.name == executable }) else {
+            // If the exectuable is explicitly specified, search through all products.
+            guard let executableProduct = plan.graph.allProducts.first(where: {
+                $0.type == .executable && $0.name == executable
+            }) else {
                 throw RunError.executableNotFound(executable)
             }
-
```
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftRunTool.swift]: nit: avoid whitespace diffs

File: Sources/Commands/SwiftRunTool.swift [hunk 3]
```diff
@@ -126,27 +126,30 @@ public class SwiftRunTool: SwiftTool<RunToolOptions> {
...
+                throw RunError.noExecutableFound
+            }
+
             // Only implicitly deduce the executable if it is the only one.
-            guard executableProducts.count == 1 else {
-                throw RunError.multipleExecutables(executableProducts.map({ $0.name }))
+            guard rootExecutables.count == 1 else {
+                throw RunError.multipleExecutables(rootExecutables.map({ $0.name }))
             }
-
```
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftRunTool.swift]: nit: avoid whitespace diffs

File: Tests/CommandsTests/BuildToolTests.swift [hunk 4]
```diff
@@ -111,11 +111,89 @@ final class BuildToolTests: XCTestCase {
...
+                let aPath = path.appending(component: "A")
+                let output = try execute(["--product", "cexec"], packagePath: aPath)
+                XCTAssert(output.contains("Compile") && output.contains("CTarget"))
+                XCTAssert(output.contains("Linking") && output.contains("cexec"))
+                XCTAssert(!output.contains("aexec"))
+                XCTAssert(!output.contains("ATarget"))
+                XCTAssert(!output.contains("BLibrary"))
+                XCTAssert(!output.contains("bexec"))
+                XCTAssert(!output.contains("BTarget1"))
+                XCTAssert(!output.contains("BTarget2"))
```
Comment (aciidb0mb3r) [inline Tests/CommandsTests/BuildToolTests.swift]: I'll land some APIs for string pattern matching. That should help a lot in this type of asserts.

File: Sources/Build/BuildPlan.swift (hunk 1 above)
Comment (hartbit) [inline Sources/Build/BuildPlan.swift]: Fixed now.

File: Sources/Build/llbuild.swift [hunk 5]
```diff
@@ -80,18 +83,24 @@ public struct LLBuildManifestGenerator {
...
-            switch buildTarget {
-            case .swift(let target):
-                targets.append(createSwiftCompileTarget(target), isTest: target.isTestTarget)
-            case .clang(let target):
-                targets.append(createClangCompileTarget(target), isTest: target.isTestTarget)
+        for (target, description) in plan.targetMap {
+            switch description {
+            case .swift(let description):
+                targets.append(createSwiftCompileTarget(description),
+                    isReachable: plan.graph.targets.contains(target),
```
Comment (aciidb0mb3r) [inline Sources/Build/llbuild.swift]: This will be slow. We can add a method in package graph which internally uses a set for fast lookup.

Comment (aciidb0mb3r) [inline Sources/Build/llbuild.swift]: We should also add some comments here about the `isReachable` and why it is needed.

File: Sources/Build/llbuild.swift [hunk 6]
```diff
@@ -51,7 +51,7 @@ public struct LLBuildManifestGenerator {
         private var otherTargets: [Target] = []
 
         /// Append a command.
-        mutating func append(_ target: Target, isTest: Bool) {
+        mutating func append(_ target: Target, isReachable: Bool, isTest: Bool) {
```
Comment (aciidb0mb3r) [inline Sources/Build/llbuild.swift]: How about `buildByDefault`?

File: Sources/Commands/SwiftRunTool.swift [hunk 7]
```diff
@@ -126,27 +126,30 @@ public class SwiftRunTool: SwiftTool<RunToolOptions> {
...
+            guard let executableProduct = plan.graph.allProducts.first(where: {
+                $0.type == .executable && $0.name == executable
+            }) else {
                 throw RunError.executableNotFound(executable)
             }
             
             return executableProduct
         } else {
+            // If the executably is implicit, search through root products.
+            let rootExecutables = plan.graph.rootPackages.flatMap({ $0.products }).filter({ $0.type == .executable })
```
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftRunTool.swift]: nit: `matchingRootExecutables`

Comment (hartbit) [inline Sources/Commands/SwiftRunTool.swift]: I don't mind, but is it really more explicit? The predicate we are matching against is already in the name: it's the list of "root executables".

File: Tests/CommandsTests/BuildToolTests.swift [hunk 8]
```diff
@@ -111,11 +113,128 @@ final class BuildToolTests: XCTestCase {
...
+
+        XCTAssertEqual(Set(graph.products.map({ $0.name })), ["aexec", "ATarget", "BLibrary"])
+        XCTAssertEqual(Set(graph.targets.map({ $0.name })), ["ATarget", "BTarget1"])
+        XCTAssertEqual(Set(graph.allProducts.map({ $0.name })), ["aexec", "ATarget", "BLibrary", "bexec", "cexec"])
+        XCTAssertEqual(Set(graph.allTargets.map({ $0.name })), ["ATarget", "BTarget1", "BTarget2", "CTarget"])
+
+        let plan = try BuildPlan(buildParameters: mockBuildParameters(), graph: graph)
+
+        XCTAssertEqual(Set(plan.buildProducts.map({ $0.product.name })), ["aexec", "ATarget", "BLibrary", "bexec", "cexec"])
+        XCTAssertEqual(Set(plan.targets.map({
```
Comment (aciidb0mb3r) [inline Tests/CommandsTests/BuildToolTests.swift]: We have `BuildPlanResult` to assert the build plan in an easier way.

File: Tests/CommandsTests/BuildToolTests.swift [hunk 9]
```diff
@@ -111,11 +113,128 @@ final class BuildToolTests: XCTestCase {
         }
     }
 
+    func testNonReachableProductsAndTargets() throws {
```
Comment (aciidb0mb3r) [inline Tests/CommandsTests/BuildToolTests.swift]: This test should be in `BuildPlanTests``

File: Tests/CommandsTests/BuildToolTests.swift [hunk 10]
```diff
@@ -12,7 +12,9 @@ import XCTest
 
 import TestSupport
 import Basic
+import Build
 import Commands
+import PackageDescription4
```
Comment (aciidb0mb3r) [inline Tests/CommandsTests/BuildToolTests.swift]: Are these imports required?

File: Tests/BuildTests/BuildPlanTests.swift [hunk 11]
```diff
@@ -19,38 +19,8 @@ import PackageModel
 import PackageDescription
 import PackageDescription4
 
-private struct MockToolchain: Toolchain {
```
Comment (aciidb0mb3r) [inline Tests/BuildTests/BuildPlanTests.swift]: This doesn't need to be moved now, right?

File: Tests/CommandsTests/BuildToolTests.swift (hunk 10 above)
Comment (hartbit) [inline Tests/CommandsTests/BuildToolTests.swift]: Nope, nice catch!

File: Tests/BuildTests/BuildPlanTests.swift (hunk 11 above)
Comment (hartbit) [inline Tests/BuildTests/BuildPlanTests.swift]: Correct!

**Comment by hartbit** [comment]
@swift-ci please smoke test

**Comment by hartbit** [comment]
@swift-ci please smoke test

**Pull request merged**